src/modules/payload_deliverer/payload_deliverer.cpp

```diff
@@ -151,22 +150,86 @@ void PayloadDeliverer::handle_vehicle_command(const hrt_abstime &now,  const veh
 
 	// Process DO_GRIPPER vehicle command
 	if (vehicle_command->command == vehicle_command_s::VEHICLE_CMD_DO_GRIPPER) {
-		// If we received Gripper command and gripper isn't valid, warn the user
 		if (!_gripper.is_valid()) {
 			PX4_WARN("Gripper instance not valid but DO_GRIPPER vehicle command was received. Gripper won't work!");
+			send_gripper_vehicle_command_ack(now, vehicle_command_ack_s::VEHICLE_CMD_RESULT_FAILED, vehicle_command->source_system,
+							 vehicle_command->source_component);
 			return;
 		}
 
-		const int32_t gripper_action = *(int32_t *)&vehicle_command->param2; // Convert the action to integer
+		const int32_t gripper_action = (int32_t)roundf(vehicle_command->param2);
+
+		// Flag to indicate if we can process the new gripper command
+		bool process_current_gripper_cmd{false};
+
+		// We are currently in the middle of executing a previous vehicle command. Handle the conflicts
+		if (_cur_vcmd_gripper_action != GRIPPER_ACTION_NONE) {
 
-		switch (gripper_action) {
-		case vehicle_command_s::GRIPPER_ACTION_GRAB:
-			_gripper.grab();
-			break;
+			if (gripper_action != _cur_vcmd_gripper_action) {
+				// If different action is commanded, cancel the previous vehicle command
+				send_gripper_vehicle_command_ack(now, vehicle_command_ack_s::VEHICLE_CMD_RESULT_CANCELLED, _cur_vcmd_target_system,
+								 _cur_vcmd_target_component);
 
-		case vehicle_command_s::GRIPPER_ACTION_RELEASE:
-			_gripper.release();
-			break;
+				// NOTE: Does this make sense when the same entity sends two conflicting commands?
+				// For the previous command, the external entity will received CANCELLED result, but then
+				// would receive IN_PROGRESS for the new command. As the COMMAND_CANCEL isn't defined properly
+				// in MAVLink yet(https://mavlink.io/en/messages/common.html#COMMAND_CANCEL), this can lead to
+				// Ground stations to thinking it's new command has been canceled instead of the previous one!
+
+				process_current_gripper_cmd = true;
```

Comment:
Regarding this case, I have referred to the MAVLink doc: https://mavlink.io/en/services/command.html#long_running_commands

It states:
```
Only one instance of a particular long running command can execute at a time; to restart a long running operation (i.e. with new parameters) it must first be cancelled!
```

So if a user wants to send for example: Gripper Close/Open/Close commands consequently (one after another, interval of around 0.2 seconds for example), GCS must send the [COMMAND_CANCEL](https://mavlink.io/en/messages/common.html#COMMAND_CANCEL) for canceling the previous command before sending the new command.

However, as stated in the COMMAND_CANCEL documentation, it is `WORK IN PROGRESS: Do not use in stable production environments (it may change).`

Which indicates that GCS may not follow this COMMAND_CANCEL procedure as documented in COMMAND_LONG, and hence the condition I mention in the comment may occur (although testing has showed that no errors pop up to user side from GCS).

I think further discussion on implementation of COMMAND_CANCEL would be needed (as I believe currently PX4 doesn't implement this part at all, as indicated by [this comment](https://github.com/mavlink/qgroundcontrol/blob/aa666d468f0dedc45a89d68250212e253ee668cb/src/AutoPilotPlugins/PX4/SensorsComponentController.cc#L491-L496) in QGC and [here](https://github.com/mavlink/qgroundcontrol/blob/aa666d468f0dedc45a89d68250212e253ee668cb/src/Vehicle/Vehicle.cc#L3459-L3471)& search in PX4 codebase)

FYI: @dagar @bkueng @ThomasDebrunner @hamishwillee